pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -212,9 +205,19 @@ func (gc *GarbageCollector) Sync(discoveryClient discovery.DiscoveryInterface, p
 			utilruntime.HandleError(fmt.Errorf("failed to sync resource monitors: %v", err))
 			return
 		}
+		// TODO: WaitForCacheSync can block forever during normal operation. Could
```

Comment:
When will it block forever?

Reply:
if newResources contains GVKs that are being removed, and are gone by the time we get here, I think blocks until the informers for newResources are all synced, which will never happen successfully for the now-missing GVKs.

1. add new resource (CRD, add-on APIService, etc)
2. GC GetDeletableResources sees new resource, notices something has changed
3. delete new resource
4. GC resyncMonitors sets up new monitors/informers for newResources, including for the now-gone resource
5. GC WaitForCacheSync waits until the informers are all synced, which will never succeed for the now-gone resource

Reply:
Shall we plumb a flag into the reflector to instruct it to not retry listing if the error is 404?

Reply:
I don't know... requires more thought. @ironcladlou can you open an issue to track this to make sure it stays high on our radar?

Reply:
Is there an open issue for this? Based on the logs, I think it is the cause of #60037

Reply:
@liggitt @ironcladlou I couldn't find the issue tracking this race

I am wondering if something like https://github.com/kubernetes/kubernetes/pull/61057 will be enough to fix the race?

Reply:
@jennybuckley looks like I neglected to create the followup issue for this. I've been slammed the past couple weeks with things unrelated to Kube but I'll try to take a look at https://github.com/kubernetes/kubernetes/pull/61057 today. Thank you!